File: Python/ByRound/2038/2038_F_Alternative_Platforms.py
# Problem: CF 2038 F - Alternative Platforms
# https://codeforces.com/contest/2038/problem/F

"""
Algorithm: Fast Fourier Transform (FFT) based approach using Number Theoretic Transform (NTT)
to compute the average user experience over all subsets of size k.
Time Complexity: O(n log n)
Space Complexity: O(n)

This solution computes the average user experience for each subset size k from 1 to n.
The function E(b1,...,bk) = max(min(v[b1], ..., v[bk]), min(r[b1], ..., r[bk]))
is evaluated over all possible k-element subsets.
Using polynomial multiplication via NTT, we derive the contribution of each blogger
to the final answer for all k-subsets and then average over C(n, k).
"""

import sys

mod = 998244353
G = 3
GI = (mod + 1) // 3


def qpow(a, b):
    # Fast exponentiation with modulo
    c = 1
    while b > 0:
        if b & 1:
            c = c * a % mod
        a = a * a % mod
        b >>= 1
    return c


def init(n):
    # Initialize bit-reversal permutation for NTT
    L = 1
    while L <= n:
        L <<= 1
    lb = [0] * L
    for i in range(L):
        lb[i] = (lb[i >> 1] >> 1) | ((i & 1) and (L >> 1) or 0)
    return L, lb


def NTT(a, fl, L, lb):
    # Number Theoretic Transform for polynomial multiplication
    for i in range(L):
        if i < lb[i]:
            a[i], a[lb[i]] = a[lb[i]], a[i]
    o = 1
    while o < L:
        Wn = qpow(G if fl else GI, (mod - 1) // (o << 1))
        for i in range(0, L, o << 1):
            w = 1
            for j in range(o):
                x = a[i + j]
                y = a[i + j + o] * w % mod
                a[i + j] = (x + y) % mod
                a[i + j + o] = (x - y) % mod
                w = w * Wn % mod
        o <<= 1
    if not fl:
        inv_L = qpow(L, mod - 2)
        for i in range(L):
            a[i] = a[i] * inv_L % mod


def C(a, b, jc, ij):
    # Compute combination C(a, b) mod mod using precomputed factorials
    return jc[a] * ij[b] % mod * ij[a - b] % mod


def Sol(c, n, tp, jc, ij, ans):
    # Polynomial-based calculation of contributions from sorted values of c
    c_sorted = sorted(c[1 : n + 1])
    c_sorted.insert(0, 0)
    max_len = 1
    while max_len <= 2 * n:
        max_len <<= 1
    e = [0] * max_len
    f = [0] * max_len
    for i in range(1, n + 1):
        e[n - i + 1] = jc[n - i + 1] * (c_sorted[i] - c_sorted[i - 1]) % mod
    for i in range(n // 2 + 1):
        e[i], e[n - i] = e[n - i], e[i]
    for i in range(n + 1):
        f[i] = ij[i]
    L, lb_arr = init(2 * n)
    NTT(e, 1, L, lb_arr)
    NTT(f, 1, L, lb_arr)
    for i in range(L):
        e[i] = e[i] * f[i] % mod
    NTT(e, 0, L, lb_arr)
    for i in range(n + 1):
        ans[i] = (ans[i] + tp * ij[i] * e[n - i]) % mod


n = int(sys.stdin.readline())
a = list(map(int, sys.stdin.readline().split()))
b = list(map(int, sys.stdin.readline().split()))
jc = [1] * (n + 1)
ij = [1] * (n + 1)
for i in range(1, n + 1):
    jc[i] = i * jc[i - 1] % mod
    ij[i] = qpow(jc[i], mod - 2)
a = [0] + a
b = [0] + b
c = [0] * (n + 1)
ans = [0] * (n + 1)
for i in range(1, n + 1):
    c[i] = a[i]
Sol(c, n, 1, jc, ij, ans)
for i in range(1, n + 1):
    c[i] = b[i]
Sol(c, n, 1, jc, ij, ans)
for i in range(1, n + 1):
    c[i] = min(a[i], b[i])
Sol(c, n, -1, jc, ij, ans)
result = []
for i in range(1, n + 1):
    inv_C = qpow(C(n, i, jc, ij), mod - 2)
    res = (ans[i] * inv_C % mod + mod) % mod
    result.append(str(res))
print(" ".join(result))


# https://github.com/VaHiX/CodeForces/

File: CPP/ByTier/F/2038_F_Alternative_Platforms.cpp
// Problem: CF 2038 F - Alternative Platforms
// https://codeforces.com/contest/2038/problem/F

/*
 * Problem: F. Alternative Platforms
 * 
 * Purpose: Calculate average user experience for all subsets of bloggers of size k,
 *          where experience is the minimum of max of videos on each platform for
 *          selected bloggers.
 * 
 * Algorithms/Techniques:
 * - Fast Fourier Transform (FFT) for polynomial multiplication
 * - Inclusion-exclusion principle
 * - Prefix sums for cumulative counting
 * - Modular arithmetic for handling large numbers
 * 
 * Time Complexity: O(n log n)
 * Space Complexity: O(n)
 *
 * Key Steps:
 * 1. Count frequency of min(v[i], r[i]) values for each blogger
 * 2. Use suffix sums to compute cumulative counts
 * 3. Build generating functions for inclusion-exclusion
 * 4. Multiply polynomials using FFT
 * 5. Extract coefficients for final answer
 * 
 * The solution uses FFT to efficiently compute convolution of polynomials,
 * which helps in calculating the inclusion-exclusion sums.
 */

#include <algorithm>
#include <iostream>
#include <utility>
#include <vector>

using namespace std;
#define endl '\n'
#define ll long long
#define sz(x) (int)x.size()
#define all(x) x.begin(), x.end()
#define FAST                                                                   \
  ios::sync_with_stdio(0);                                                     \
  cin.tie(0);
#define getunique(v)                                                           \
  {                                                                            \
    sort(v.begin(), v.end());                                                  \
    v.erase(unique(v.begin(), v.end()), v.end());                              \
  }
#define random                                                                 \
  mt19937_64 rng(chrono::steady_clock::now().time_since_epoch().count());      \
  mt19937 rnd(time(0));
const int INF = 1e9;
const int MX = 1e6 + 9;
const int MOD = 998244353;
int norm(int x) {
  if (x < 0) {
    x += MOD;
  }
  if (x >= MOD) {
    x -= MOD;
  }
  return x;
}
int add(int a, int b) { return a + b < MOD ? a + b : a + b - MOD; }
int mul(int a, int b) { return a * (long long)b % MOD; }
int sub(int a, int b) { return a >= b ? a - b : a - b + MOD; }
int power(int a, long long p = MOD - 2) {
  int ans = 1;
  for (; p > 0; p >>= 1, a = mul(a, a)) {
    if (p & 1) {
      ans = mul(ans, a);
    }
  }
  return ans;
}
vector<int> rev;
vector<int> roots{0, 1};
void dft(vector<int> &a) {
  int n = (int)a.size();
  if (n == 1)
    return;
  if (int(rev.size()) != n) {
    int k = __builtin_ctz(n) - 1;
    rev.resize(n);
    for (int i = 0; i < n; i++) {
      rev[i] = rev[i >> 1] >> 1 | (i & 1) << k;
    }
  }
  for (int i = 0; i < n; i++) {
    if (rev[i] < i) {
      swap(a[i], a[rev[i]]);
    }
  }
  if (int(roots.size()) < n) {
    int k = __builtin_ctz(roots.size());
    roots.resize(n);
    while ((1 << k) < n) {
      int e = power(int(3), (MOD - 1) >> (k + 1));
      for (int i = 1 << (k - 1); i < (1 << k); i++) {
        roots[i << 1] = roots[i];
        roots[i << 1 | 1] = mul(roots[i], e);
      }
      k++;
    }
  }
  for (int k = 1; k < n; k <<= 1) {
    for (int i = 0; i < n; i += k << 1) {
      for (int j = 0; j < k; ++j) {
        int u = a[i + j], v = mul(a[i + j + k], roots[k + j]);
        a[i + j] = add(u, v);
        a[i + j + k] = sub(u, v);
      }
    }
  }
}
void idft(vector<int> &a) {
  int n = (int)a.size();
  reverse(a.begin() + 1, a.end());
  dft(a);
  int inverse = norm((1 - MOD) / n);
  for (auto &v : a)
    v = mul(v, inverse);
}
vector<int> mul(vector<int> a, vector<int> b) {
  if (a.empty() || b.empty()) {
    return {};
  }
  int sz = 1, tot = int(a.size() + b.size() - 1);
  while (sz < tot)
    sz <<= 1;
  a.resize(sz), b.resize(sz);
  dft(a);
  dft(b);
  for (int i = 0; i < sz; ++i) {
    a[i] = mul(a[i], b[i]);
  }
  idft(a);
  a.resize(tot);
  return a;
}
vector<int> fact(MX), inv(MX), factin(MX);
void init() {
  fact[0] = fact[1] = inv[1] = factin[0] = factin[1] = 1;
  for (int i = 2; i < MX; i++) {
    fact[i] = mul(fact[i - 1], i);
    inv[i] = mul(sub(MOD, MOD / i), inv[MOD % i]);
    factin[i] = mul(factin[i - 1], inv[i]);
  }
}
void solve() {
  int n;
  cin >> n;
  vector<int> a(n);
  for (int i = 0; i < n; i++) {
    cin >> a[i];
  }
  vector<int> b(n);
  for (int i = 0; i < n; i++) {
    cin >> b[i];
  }
  vector<int> ga(MX), gb(MX), gab(MX);
  // Count frequencies of values
  for (int i = 0; i < n; i++) {
    ++ga[a[i]];
    ++gb[b[i]];
    ++gab[min(a[i], b[i])];
  }
  // Suffix sum to get cumulative counts
  for (int i = MX - 1; i > 0; i--) {
    ga[i - 1] += ga[i];
    gb[i - 1] += gb[i];
    gab[i - 1] += gab[i];
  }
  // Build generating function with inclusion-exclusion
  vector<int> A(n + 1), B(n + 1);
  for (int i = 1; i < MX; i++) {
    int cnt = 0;
    cnt = ga[i];
    A[cnt] = add(A[cnt], 1);  // Add bloggers contributing to this count from first platform
    cnt = gb[i];
    A[cnt] = add(A[cnt], 1);  // Add bloggers contributing to this count from second platform
    cnt = gab[i];
    A[cnt] = sub(A[cnt], 1);  // Subtract those that are in both (inclusion-exclusion)
  }
  for (int cnt = 0; cnt <= n; cnt++) {
    A[cnt] = mul(A[cnt], fact[cnt]);  // Multiply by factorial for arrangement
    B[n - cnt] = factin[cnt];         // Inverse factorial for normalization
  }
  // Polynomial multiplication using FFT
  auto tot = mul(A, B);
  vector<int> ans(n + 1);
  for (int k = 1; k <= n; k++) {
    if (n + k >= sz(tot)) {
      break;
    }
    ans[k] = tot[n + k];              // Extract coefficient for k-size subset
    ans[k] = mul(ans[k], fact[n - k]); // Correct for arrangement factor
    ans[k] = mul(ans[k], factin[n]);   // Divide by total arrangements
  }
  for (int i = 1; i <= n; i++) {
    cout << ans[i] << " ";
  }
}
int main() {
  FAST;
  init();
  int t = 1;
  while (t--) {
    solve();
  }
}


// https://github.com/VaHiX/CodeForces/